src/utils.jl

```diff
@@ -28,42 +28,65 @@ function get_args(func_def::Dict)
   arg_list, kwarg_list, arg_dict
 end
 
+function forward_args(func_def)
+  args = []
+  map!(func_def[:args], func_def[:args]) do arg
+    name, type, splat, default = splitarg(arg)
+    name = something(name, gensym())
+    if splat
+      push!(args, :($name...))
+    else
+      push!(args, name)
+    end
+    combinearg(name, type, splat, default)
+  end
+  kwargs = []
+  for arg in func_def[:kwargs]
+    name, type, splat, default = splitarg(arg)
+    if splat
+      push!(kwargs, :($name...))
+    else
+      push!(kwargs, name)
+    end
+  end
+  args, kwargs
+end
+
+const unused = (Symbol("#temp#"), Symbol("_"), Symbol(""), Symbol("#unused#"), Symbol("#self#"))
+
 """
 Function returning the slots of a function definition
 """
-function get_slots(func_def::Dict, args::Dict{Symbol, Any}, mod::Module) :: Dict{Symbol, Any}
+function get_slots(func_def::Dict, args::Dict{Symbol, Any}, mod::Module)
   slots = Dict{Symbol, Any}()
   func_def[:name] = gensym()
   func_def[:args] = (func_def[:args]..., func_def[:kwargs]...)
   func_def[:kwargs] = []
-  body = func_def[:body]
+  # replace yield with inference barrier
   func_def[:body] = postwalk(transform_yield, func_def[:body])
+  # collect items to skip
   nosaves = Set{Symbol}()
   func_def[:body] = postwalk(x->transform_nosave(x, nosaves), func_def[:body])
+  # eval function
   func_expr = combinedef(func_def) |> flatten
   @eval(mod, @noinline $func_expr)
+  # get typed code
   codeinfos = @eval(mod, code_typed($(func_def[:name]), Tuple; optimize=false))
+  # extract slot names and types
   for codeinfo in codeinfos
     for (name, type) in collect(zip(codeinfo.first.slotnames, codeinfo.first.slottypes))
-      name ∉ nosaves && (slots[name] = type)
+      name ∉ nosaves && name ∉ unused && (slots[name] = Union{type, get(slots, name, Union{})})
     end
   end
-  for (argname, argtype) in args
-    slots[argname] = argtype
-  end
+  # remove `catch exc` statements
   postwalk(x->remove_catch_exc(x, slots), func_def[:body])
-  postwalk(x->make_arg_any(x, slots), body)
```

Comment:
Is `make_arg_any` used by anything else at this point? If it is not used, maybe we should delete it?

Why is `make_arg_any` not necessary here anymore? Do you know what its original purpose was?

Reply:
`inferencebarrier` takes care of this now. The original purpose was that `val = @yield ret` would expand to `val = ret` and infer the type of `val` as the type of `ret` rather than Any. `make_arg_any would manually pass through and turn them into Any, while now we just tell the compiler to not infer through the barrier.